display_list/display_list_utils.cc

```diff
@@ -258,11 +258,143 @@ void ClipBoundsDispatchHelper::reset(const SkRect* cull_rect) {
   }
 }
 
+void RectBoundsAccumulator::accumulate(const SkRect& r) {
+  if (r.fLeft < r.fRight && r.fTop < r.fBottom) {
+    rect_.accumulate(r.fLeft, r.fTop);
+    rect_.accumulate(r.fRight, r.fBottom);
+  }
+}
+
+void RectBoundsAccumulator::save() {
+  saved_rects_.emplace_back(rect_);
+  rect_ = AccumulationRect();
+}
+void RectBoundsAccumulator::restore(const SkRect* clip) {
+  if (!saved_rects_.empty()) {
+    SkRect layer_bounds = rect_.bounds();
+    pop_and_accumulate(layer_bounds, clip);
+  }
+}
+bool RectBoundsAccumulator::restore(
+    std::function<bool(const SkRect&, SkRect&)> mapper,
+    const SkRect* clip) {
+  bool success = true;
+  if (!saved_rects_.empty()) {
+    SkRect layer_bounds = rect_.bounds();
+    success = mapper(layer_bounds, layer_bounds);
+    pop_and_accumulate(layer_bounds, clip);
+  }
+  return success;
+}
+void RectBoundsAccumulator::pop_and_accumulate(SkRect& layer_bounds,
+                                               const SkRect* clip) {
+  FML_DCHECK(!saved_rects_.empty());
+
+  rect_ = saved_rects_.back();
+  saved_rects_.pop_back();
+
+  if (clip == nullptr || layer_bounds.intersect(*clip)) {
+    accumulate(layer_bounds);
+  }
+}
+
+RectBoundsAccumulator::AccumulationRect::AccumulationRect() {
+  min_x_ = std::numeric_limits<SkScalar>::infinity();
+  min_y_ = std::numeric_limits<SkScalar>::infinity();
+  max_x_ = -std::numeric_limits<SkScalar>::infinity();
+  max_y_ = -std::numeric_limits<SkScalar>::infinity();
+}
+void RectBoundsAccumulator::AccumulationRect::accumulate(SkScalar x,
+                                                         SkScalar y) {
+  if (min_x_ > x) {
+    min_x_ = x;
+  }
+  if (min_y_ > y) {
+    min_y_ = y;
+  }
+  if (max_x_ < x) {
+    max_x_ = x;
+  }
+  if (max_y_ < y) {
+    max_y_ = y;
+  }
+}
+SkRect RectBoundsAccumulator::AccumulationRect::bounds() const {
+  return (max_x_ >= min_x_ && max_y_ >= min_y_)
+             ? SkRect::MakeLTRB(min_x_, min_y_, max_x_, max_y_)
+             : SkRect::MakeEmpty();
+}
+
+void RTreeBoundsAccumulator::accumulate(const SkRect& r) {
+  if (r.fLeft < r.fRight && r.fTop < r.fBottom) {
+    rects_.push_back(r);
+  }
+}
+bool RTreeBoundsAccumulator::is_empty() const {
+  return rects_.empty();
+}
+bool RTreeBoundsAccumulator::is_not_empty() const {
+  return !rects_.empty();
+}
+void RTreeBoundsAccumulator::save() {
+  saved_offsets_.push_back(rects_.size());
+}
+void RTreeBoundsAccumulator::restore(const SkRect* clip) {
+  if (saved_offsets_.empty()) {
+    return;
+  }
+
+  size_t previous_size = saved_offsets_.back();
+  saved_offsets_.pop_back();
+
+  if (clip) {
+    for (size_t i = previous_size; i < saved_offsets_.size(); i++) {
```

Comment:
Here and another similar place, `rects_` should be used instead of `saved_offsets_`, right?
Should we add a test for this? 
for example
```
TEST(DisplayList, RTreeOfClipRectScene) {
  DisplayListBuilder builder;
  builder.drawRect({10, 10, 20, 20});
  builder.save();
  builder.clipRect({50, 50, 60, 60}, SkClipOp::kIntersect, false);
  builder.drawRect({30, 30, 100, 100});
  builder.restore();
  auto display_list = builder.Build();
  auto rtree = display_list->rtree();
  std::vector<SkRect> rects = {
      {10, 10, 20, 20},
      {50, 50, 60, 60},
  };

  // Missing all drawRect calls
  test_rtree(rtree, {5, 5, 10, 10}, rects, {});
  test_rtree(rtree, {20, 20, 25, 25}, rects, {});
  test_rtree(rtree, {45, 45, 50, 50}, rects, {});
  test_rtree(rtree, {60, 60, 65, 65}, rects, {});

  // Hitting just 1 of the drawRects
  test_rtree(rtree, {5, 5, 11, 11}, rects, {0});
  test_rtree(rtree, {19, 19, 25, 25}, rects, {0});
  test_rtree(rtree, {45, 45, 51, 51}, rects, {1});
  test_rtree(rtree, {59, 59, 65, 65}, rects, {1});

  // Hitting both drawRect calls
  test_rtree(rtree, {19, 19, 51, 51}, rects, {0, 1});
}
```

Reply:
Ah, good catch, but I think the bigger picture is that we don't need to apply the clip at all in this case. It is needed for the restore-with-filter case because the filter might expand the bounds and they need to be *RE*-clipped, but for a regular save, the clip isn't an issue at all because it was already applied during the accumulations.

Also, the clip being applied here is not the clip you specify with clipRect, it is the default cull_rect from the builder. This issue would have happened even without the call to clipRect, as long as there was a save/restore.

Reply:
Yes, it fails with just a save/restore.

Reply:
Done, plus removed clip handling from the non-filtered restore.